packages/flutter_tools/lib/src/android/android_builder.dart

```diff
@@ -0,0 +1,169 @@
+              'for details on how to upgrade the project.'
+      );
+    }
+    if (!project.manifest.isModule && !project.manifest.isPlugin) {
+      throwToolExit('AARs can only be built for plugin or module projects.');
+    }
+    // Validate that we can find an Android SDK.
+    if (androidSdk == null) {
+      throwToolExit('No Android SDK found. Try setting the `ANDROID_SDK_ROOT` environment variable.');
+    }
+    await buildGradleAar(
+      project: project,
+      androidBuildInfo: androidBuildInfo,
+      target: target,
+      outputDir: outputDir,
+    );
+    androidSdk.reinitialize();
+  }
+
+  /// Builds the APK.
+  @override
+  Future<void> buildApk({
+    @required FlutterProject project,
+    @required AndroidBuildInfo androidBuildInfo,
+    @required String target,
+  }) async {
+     if (!project.android.isUsingGradle) {
+      throwToolExit(
+          'The build process for Android has changed, and the current project configuration '
+              'is no longer valid. Please consult\n\n'
+              '  https://github.com/flutter/flutter/wiki/Upgrading-Flutter-projects-to-build-with-gradle\n\n'
+              'for details on how to upgrade the project.'
+      );
+    }
+    // Validate that we can find an android sdk.
+    if (androidSdk == null) {
+      throwToolExit('No Android SDK found. Try setting the ANDROID_SDK_ROOT environment variable.');
+    }
+    await buildGradleProject(
+      project: project,
+      androidBuildInfo: androidBuildInfo,
+      target: target,
+      isBuildingBundle: false,
+    );
+    androidSdk.reinitialize();
+  }
+
+  /// Builds the App Bundle.
+  @override
+  Future<void> buildAab({
+    @required FlutterProject project,
+    @required AndroidBuildInfo androidBuildInfo,
+    @required String target,
+  }) async {
+    if (!project.android.isUsingGradle) {
+      throwToolExit(
+          'The build process for Android has changed, and the current project configuration '
+          'is no longer valid. Please consult\n\n'
+          'https://github.com/flutter/flutter/wiki/Upgrading-Flutter-projects-to-build-with-gradle\n\n'
+          'for details on how to upgrade the project.'
+      );
+    }
+    // Validate that we can find an android sdk.
+    if (androidSdk == null) {
+      throwToolExit('No Android SDK found. Try setting the ANDROID_HOME environment variable.');
+    }
+    final List<String> validationResult = androidSdk.validateSdkWellFormed();
+    if (validationResult.isNotEmpty) {
+      for (String message in validationResult) {
+        printError(message, wrap: false);
+      }
+      throwToolExit('Try re-installing or updating your Android SDK.');
+    }
+    return buildGradleProject(
+      project: project,
+      androidBuildInfo: androidBuildInfo,
+      target: target,
+      isBuildingBundle: true,
+    );
+  }
+}
+
+/// A fake implementation of [AndroidBuilder].
+@visibleForTesting
+class FakeAndroidBuilder implements AndroidBuilder {
```

Comment:
This should probably be in a testing file

Reply:
This class is used in multiple test files, so it seems useful to keep it in the same file as the abstract class.  The linter should catch if FakeAndroidBuilder is used outside of a `*_test.dart` file.